### PR TITLE
Docs for adding new cloud provider

### PR DESCRIPTION
**Pull request by deitch**
**What this PR does / why we need it**:

Overview document for how to add a new cloud provider

**Special notes for your reviewer**:
As discussed in Slack with @rfranzke .

Note that there are document links missing, marked as `In process`, since there are documents missing. Once this is agreed as good and merged in, will add those documents as separate PRs, and then update this doc with the links.

**Release note**:

```improvement user
NONE
```

### CODE REVIEW

File: docs/development/new-cloud-provider.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# Adding Cloud Providers
```
Comment (zanetworker) [inline docs/development/new-cloud-provider.md:1]: Add space after header

File: docs/development/new-cloud-provider.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+This document provides an overview of how to integrate a new cloud provider into Gardener. Each component that requires integration has a detailed description of how to integrate it and the steps required.
+
+## Cloud Components
```
Comment (zanetworker) [inline docs/development/new-cloud-provider.md:5]: space after header

File: docs/development/new-cloud-provider.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+* Shoot: These are one more "workload" clusters, the regular Kubernetes clusters you have seen. They provide places for your workloads to run.
+* Seed: This is the "management" cluster. It provides management service for one or more Shoot cluster.
+
+These two clusters can run in the same cloud provider, but they do not need to. For example, you could run your Seed in AWS, while having one shoot in Azure, two in Google, two in Alicloud, and three in Packet.
+
+The Seed cluster deploys and manages the Shoot clusters. Importantly, for this discussion, the `etcd` data store backing each Shoot runs as workloads inside the Seed. Thus, to use the above example, the clusters in Azure, Google, Alicloud and Packet will have their worker nodes and master nodes running in those clouds, but the `etcd` clusters backing them will run as separate [deployments](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) in the Seed Kubernetes cluster on AWS.
+
+This distinction becomes important when preparing to integrate a new cloud provider.
+
+## Gardener Cloud Integration
```
Comment (zanetworker) [inline docs/development/new-cloud-provider.md:18]: space after header

File: docs/development/new-cloud-provider.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+* Create/destroy/get/list machines for the Seed
+* Backup/restore etcd for the Seed via writing files to and reading them from object storage
+
+Thus, which integrations you need for your cloud provider depend on if you want to deploy Shoot clusters to the provider, Seed or both.
+
+* Shoot Only: machine lifecycle management
+* Seed: machine lifecycle management **and** etcd backup/restore
+
+
+## Gardener API
```
Comment (zanetworker) [inline docs/development/new-cloud-provider.md:32]: same as above

File: docs/development/new-cloud-provider.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+In addition to the requirements to integrate with the cloud provider, you also need to enable the core Gardener app to receive, validate and process requests to use that cloud provider.
+
+* Expose the cloud provider to the consumers of the Gardener API, so it can be told to use that cloud provider as an option
+* Validate that API as requests come in
+* Extend Botanist to deploy machines to the cloud provider
+
+## Cloud Provider API Requirements
```
Comment (zanetworker) [inline docs/development/new-cloud-provider.md:40]: same as above

File: docs/development/new-cloud-provider.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+* Create a machine
+* Destroy a machine
+* Get information about a machine and its state
+* List machines
+
+As Gardener is written in [go](https://golang.org), the cloud provider should have a go SDK. However, if it has an API that is wrappable in go, e.g. a REST API, then you can use that to integrate.
+
+In addition, if the Seed is to run on the given provider, it also must have an API to save files to block storage and retrieve them, for etcd backup/restore.
+
+## Summary
```
Comment (zanetworker) [inline docs/development/new-cloud-provider.md:55]: same as above

File: docs/development/new-cloud-provider.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+* Create/destroy/get/list machines for the Shoot
+* Create/destroy/get/list machines for the Seed
+* Backup/restore etcd for the Seed via writing files to and reading them from object storage
+
+Thus, which integrations you need for your cloud provider depend on if you want to deploy Shoot clusters to the provider, Seed or both.
+
+* Shoot Only: machine lifecycle management
+* Seed: machine lifecycle management **and** etcd backup/restore
+
+
```
Comment (zanetworker) [inline docs/development/new-cloud-provider.md:31]: remove space

File: docs/development/new-cloud-provider.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+This distinction becomes important when preparing to integrate a new cloud provider.
+
+## Gardener Cloud Integration
+Gardener integrates with cloud providers at the following key lifecycle elements:
+
+* Create/destroy/get/list machines for the Shoot
+* Create/destroy/get/list machines for the Seed
+* Backup/restore etcd for the Seed via writing files to and reading them from object storage
+
+Thus, which integrations you need for your cloud provider depend on if you want to deploy Shoot clusters to the provider, Seed or both.
```
Comment (zanetworker) [inline docs/development/new-cloud-provider.md]: maybe replace with `Thus, the integrations you need for your cloud provider depend on whether `

**Review by zanetworker — COMMENTED** [review]
Suggested minor changes, otherwise LGTM. 

@jia-jerry  maybe you have something to add here?

File: docs/development/new-cloud-provider.md (hunk 2 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md:5]: If you insist. Not sure why though? markdown automatically eliminates extra newlines there.

File: docs/development/new-cloud-provider.md (hunk 7 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md:31]: Done.

File: docs/development/new-cloud-provider.md (hunk 8 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md]: OK

**Comment by deitch** [comment]
Changed the requested language. I can add a blank line between the various headers and text, if you want, but not sure why, since the markdown doesn't change?

Compare [this](https://github.com/gardener/gardener/blob/master/docs/development/local_setup.md) which has the blank line, and [this](https://github.com/gardener/gardener/blob/master/docs/development/process.md), which does not.

**Comment by zanetworker** [comment]
@deitch this is just a Markdown linting rule (https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md#md022---headers-should-be-surrounded-by-blank-lines). Probably the older docs are not linted correctly, in anycase, feel free to leave it out if you would like, it's a minor suggestion.

File: docs/development/new-cloud-provider.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+The Seed cluster deploys and manages the Shoot clusters. Importantly, for this discussion, the `etcd` data store backing each Shoot runs as workloads inside the Seed. Thus, to use the above example, the clusters in Azure, Google, Alicloud and Packet will have their worker nodes and master nodes running in those clouds, but the `etcd` clusters backing them will run as separate [deployments](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) in the Seed Kubernetes cluster on AWS.
+
+This distinction becomes important when preparing to integrate a new cloud provider.
+
+## Gardener Cloud Integration
+Gardener integrates with cloud providers at the following key lifecycle elements:
+
+* Create/destroy/get/list machines for the Shoot
+* Create/destroy/get/list machines for the Seed
```
Comment (rfranzke) [inline docs/development/new-cloud-provider.md]: Why this distinction? I would only list it once - technically, there is none.

File: docs/development/new-cloud-provider.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+* Backup/restore etcd for the Seed via writing files to and reading them from object storage
+
```
Comment (rfranzke) [inline docs/development/new-cloud-provider.md:26]: Please also list the infrastructure preparation, and *optionally* provider-specific DNS integration.

File: docs/development/new-cloud-provider.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+Thus, the integrations you need for your cloud provider depend on whether you want to deploy Shoot clusters to the provider, Seed or both.
+
+* Shoot Only: machine lifecycle management
+* Seed: machine lifecycle management **and** etcd backup/restore
```
Comment (rfranzke) [inline docs/development/new-cloud-provider.md]: not only those two, but also configuration of the control plane components

File: docs/development/new-cloud-provider.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+## Gardener API
+In addition to the requirements to integrate with the cloud provider, you also need to enable the core Gardener app to receive, validate and process requests to use that cloud provider.
+
+* Expose the cloud provider to the consumers of the Gardener API, so it can be told to use that cloud provider as an option
+* Validate that API as requests come in
+* Extend Botanist to deploy machines to the cloud provider
```
Comment (rfranzke) [inline docs/development/new-cloud-provider.md]: replace with "Write cloud provider specific implementation (called Botanist)"

File: docs/development/new-cloud-provider.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+## Cloud Provider API Requirements
+In order for a cloud provider to integrate with Gardener, whether for Shoot or Seed clusters, the provider must have an API to perform machine lifecycle events, specifically:
+
+* Create a machine
+* Destroy a machine
+* Get information about a machine and its state
+* List machines
+
+As Gardener is written in [go](https://golang.org), the cloud provider should have a go SDK. However, if it has an API that is wrappable in go, e.g. a REST API, then you can use that to integrate.
```
Comment (rfranzke) [inline docs/development/new-cloud-provider.md]: I don't get this block - why is this listed here? That's part of the machine controller manager, right? From Gardener PoV only a MCM docker image containing the cloud integration as well as a `MachineClass` must be delivered, there is no explicit machine handling done by Gardener.

**Review by rfranzke — CHANGES_REQUESTED** [review]

**Comment by deitch** [comment]
@zanetworker wrote:

> this is just a Markdown linting rule 

OK, that makes sense. Updating now.

**Comment by deitch** [comment]
Rebased too.

File: docs/development/new-cloud-provider.md (hunk 12 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md]: I rewrote that at least 3 times, never happy with the language. I like yours better, thanks.

Comment (deitch) [reply docs/development/new-cloud-provider.md]: This is updated

File: docs/development/new-cloud-provider.md (hunk 13 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md]: Well, yes and no. From the "gardener as a component" perspective, this is true; it only needs to have an MCM image, which can be in any language. From the "gardener as a system" perspective, I want a high-level overview of all of the requirements.

I think the issue is that the docs for "gardener as a system" are in the same repo right here as "gardener as a component". The actual _implementation_ of interaction with the cloud provider - and thus the requirement for either a go SDK or a way to call the API via go - happens in MCM. But if this doc is my entry point to adding a new cloud provider to "gardener as a system", then I want to know my pre-requisites upfront.

If there were an alternate "developing for gardener as a system" doc area, I gladly would put it there.

Does that make sense?

File: docs/development/new-cloud-provider.md (hunk 9 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md]: True, _technically_ there is one. My perspective here, though, is from someone coming to add a new cloud provider. My first question (which was my first question getting started) is: "where in its lifecycle does Gardener integrate with the cloud provider?"

The answer might be at 5 distinct points, and those may reuse the same two (or even 1) component, but the answer to my question isn't "2 components", but rather, "5 points in the lifecycle, and those 5 points reuse the same 2 components."

I am trying (more or less successfully) to answer the questions as they would have made my life easier getting started a few weeks back.

File: docs/development/new-cloud-provider.md (hunk 10 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md:26]: Can you help here? Not sure what needs to be added. My involvement with integration thus far has only been looking at MCM, etch-backup-restore, and the beginnings of API+validation+Botanist. What is needed?

File: docs/development/new-cloud-provider.md (hunk 11 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md]: Same here, can you help? What does Gardener need to interact with the cloud provider API when configuring the control plane components? Happy to add it, but unsure what needs to be added.

File: docs/development/new-cloud-provider.md (hunk 9 above)
Comment (rfranzke) [reply docs/development/new-cloud-provider.md]: Gardener does not manage machines of seeds - it just gets a kubeconfig and then uses a pre-existing seed cluster. Hence, I still find it confusing.

File: docs/development/new-cloud-provider.md (hunk 10 above)
Comment (rfranzke) [reply docs/development/new-cloud-provider.md:26]: If you take a look at the [`CloudBotanist` interface](https://github.com/gardener/gardener/blob/master/pkg/operation/cloudbotanist/types.go#L25-L60) you see some methods like `DeployInfrastructure` and `DestroyInfrastructure`.
These methods are getting called during creation of shoot - for AWS and other public clouds they create VPCs, subnetworks, routes, and security groups into which the VMs (provisioned by MCM) and LBs (provisioned by CCM/KCM) are deployed later.
Not sure what is needed for packet, but I expect sth. similar.

For DNS: Users may use their cloud-specific DNS provider when integrate into Gardener. Example: Alicloud just integrated, but they cannot use AWS Route53, Google Cloud DNS or others. Hence, they integrated Alicloud DNS as a separate step (#393). This is optional.

File: docs/development/new-cloud-provider.md (hunk 11 above)
Comment (rfranzke) [reply docs/development/new-cloud-provider.md]: Same like above - take a look at the CloudBotanist interface. The K8s control plane (apiserver, controllers, scheduler, etc.) might need specific flags based on the cloud provider they serve. Usually not much, but there might be the one or the other special flag.

File: docs/development/new-cloud-provider.md (hunk 13 above)
Comment (rfranzke) [reply docs/development/new-cloud-provider.md]: > I think the issue is that the docs for "gardener as a system" are in the same repo right here as "gardener as a component".

Yes. If you want to do that then I suggest to have a separate section/paragraph in the doc to not mix things up. Provide first information what needs to be done (ONLY from Gardener component PoV), and prior/later a system PoV.

WDYT?

File: docs/development/new-cloud-provider.md (hunk 1 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md:1]: Done

File: docs/development/new-cloud-provider.md (hunk 2 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md:5]: Done

File: docs/development/new-cloud-provider.md (hunk 4 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md:32]: Done

File: docs/development/new-cloud-provider.md (hunk 6 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md:55]: Done

File: docs/development/new-cloud-provider.md (hunk 9 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md]: Oh yes, very good point. OK, will clear that up.

File: docs/development/new-cloud-provider.md (hunk 13 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md]: OK, that makes sense. I can clarify the language here.

How does that align with your comment [above](
https://github.com/gardener/gardener/pull/647#discussion_r247038326) about the Botanist interface? Doesn't it mean we need cloud provider integration in go (or written to wrap an API) for Gardener as well?

**Comment by deitch** [comment]
@rfranzke I updated the doc per your comments (didn't get _everything_, but much of it). Please take a look, especially around the conversations that remain open.

File: docs/development/new-cloud-provider.md (hunk 13 above)
Comment (rfranzke) [reply docs/development/new-cloud-provider.md]: No, it can be anything - Terraform, Go, whatever.

Comment (deitch) [reply docs/development/new-cloud-provider.md]: I don't get it. 

* MCM: a single image, with integrations to AWS, GCE, Azure, Ali, Packet. If I want to add, e.g. DigitalOcean, I need to add it into the binary, which means go, does it not?
* Botanist: How would I implement the Botanist interface (a go construct) into Gardener if it is written in something other than go? Python or nodejs wouldn't work.

Comment (rfranzke) [reply docs/development/new-cloud-provider.md]: OK, I got confused by what you mean. Sure, everything needs to be Golang as there is no way to have the providers externally yet. :)

Comment (deitch) [reply docs/development/new-cloud-provider.md]: That is a good point. I will write something about in-tree vs out-of-tree pluggable here. Coming...

Comment (deitch) [reply docs/development/new-cloud-provider.md]: Check it out now.

**Comment by deitch** [comment]
I think we have addressed everything, but let me know...

File: docs/development/new-cloud-provider.md [hunk 14]
```diff
@@ -0,0 +1,65 @@
+# Adding Cloud Providers
+
+This document provides an overview of how to integrate a new cloud provider into Gardener. Each component that requires integration has a detailed description of how to integrate it and the steps required.
+
+## Cloud Components
+
+Gardener is composed of 2 or more Kubernetes clusters:
+
+* Shoot: These are one more "workload" clusters, the regular Kubernetes clusters you have seen. They provide places for your workloads to run.
```
Comment (rfranzke) [inline docs/development/new-cloud-provider.md]: s/These are one more "workload" clusters/These are the end-user clusters

File: docs/development/new-cloud-provider.md [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+* Seed: This is the "management" cluster. It provides management service for one or more Shoot cluster.
```
Comment (rfranzke) [inline docs/development/new-cloud-provider.md]: "Management service" is quite general, let's clarify that: It manages the control planes of multiple shoot clusters by running them as native Kubernetes workload.

File: docs/development/new-cloud-provider.md [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+
+These two clusters can run in the same cloud provider, but they do not need to. For example, you could run your Seed in AWS, while having one shoot in Azure, two in Google, two in Alicloud, and three in Packet.
+
+The Seed cluster deploys and manages the Shoot clusters. Importantly, for this discussion, the `etcd` data store backing each Shoot runs as workloads inside the Seed. Thus, to use the above example, the clusters in Azure, Google, Alicloud and Packet will have their worker nodes and master nodes running in those clouds, but the `etcd` clusters backing them will run as separate [deployments](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) in the Seed Kubernetes cluster on AWS.
+
+This distinction becomes important when preparing to integrate a new cloud provider.
```
Comment (rfranzke) [inline docs/development/new-cloud-provider.md]: s/to integrate/the integration

File: docs/development/new-cloud-provider.md [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+
+## Gardener Cloud Integration
+
+Gardener integrates with cloud providers at the following key lifecycle elements:
```
Comment (rfranzke) [inline docs/development/new-cloud-provider.md]: s/Gardener/Gardener and its related components

File: docs/development/new-cloud-provider.md [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+
+* Create/destroy/get/list machines for the Shoot
+* Create/destroy/get/list infrastructure components for the Shoot, e.g. VPCs, subnets, routes, etc.
+* Create/destroy/get/list DNS zones and entries for the Shoot
+* Backup/restore etcd for the Seed via writing files to and reading them from object storage
+
+Thus, the integrations you need for your cloud provider depend on whether you want to deploy Shoot clusters to the provider, Seed or both.
+
+* Shoot Only: machine lifecycle management, infrastructure, DNS
```
Comment (rfranzke) [inline docs/development/new-cloud-provider.md]: Mention that DNS is truly optional!

**Review by rfranzke — COMMENTED** [review]
@deitch just a few minor things, after that we can merge this first version! Thank's for your efforts!

**Comment by deitch** [comment]
@rfranzke wrote:

> after that we can merge this first version! Thank's for your efforts!

Quite welcome... but don't think it is over! One this is merged in, we have to go after the docs on some of the details, and then I finally can finish the packet driver. :-)

File: docs/development/new-cloud-provider.md (hunk 14 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md]: OK

**Comment by rfranzke** [comment]
Sure :) However, let's not invest TOO much as some of the things (most notably, the Botanist part) changes with https://github.com/gardener/gardener/blob/master/docs/proposals/01-extensibility.md. We are currently working towards the implementation of this proposal, so let's focus more on the external parts (MCM, etcd-backup-restore) first.

File: docs/development/new-cloud-provider.md (hunk 17 above)
Comment (deitch) [reply docs/development/new-cloud-provider.md]: OK.

I stand by the earlier comment, that I think calling both the system as a whole "Gardener" and the core binary "Gardener" gets confusing.

**Comment by deitch** [comment]
Updated (and rebased, of course).

**Comment by deitch** [comment]
> However, let's not invest TOO much as some of the things (most notably, the Botanist part) changes with https://github.com/gardener/gardener/blob/master/docs/proposals/01-extensibility.md. We are currently working towards the implementation of this proposal, so let's focus more on the external parts (MCM, etcd-backup-restore) first

Sounds good. Once this is in, we can jump back to Slack to discuss which parts to doc next.

**Review by rfranzke — APPROVED** [review]
/lgtm, thanks!

**Pull request merged**